abra-lib/client.go

```diff
@@ -214,6 +214,81 @@ func NewWithGuid(guid string) (*Client, error) {
 	return client, nil
 }
 
+type numberType int
+
+const (
+	numberTypeNone numberType = iota
+	numberTypeABN
+	numberTypeACN
+)
+
+// Search attempts to intelligently work out what you're looking for
+func (c *Client) Search(text string) (*BusinessEntity, error) {
+	zzz, res := entityNumberFromString(text)
+
+	switch res {
+	case numberTypeNone:
+		return nil, fmt.Errorf("Not implemented")
+	case numberTypeABN:
+		return c.SearchByABN(zzz, false)
+	case numberTypeACN:
+		return c.SearchByACN(zzz, false)
+	}
+
+	return nil, fmt.Errorf("Unexpectedly reached an unreachable piece of code")
+}
+
+func entityNumberFromString(text string) (string, numberType) {
+	stringLen := len(text)
+	for i := 0; i < stringLen; i++ {
+		if charCheck(text[i]) != charCheckNumber {
+			continue
+		}
+
+		result := []byte{text[i]}
+		// get numbers until they stop
+		i++
+		for ; i < stringLen; i++ {
+			switch charCheck(text[i]) {
+			case charCheckWhitespace:
+				continue
+			case charCheckNumber:
+				result = append(result, text[i])
+			case charCheckNotNumber:
+				break
+			}
+		}
+
+		// now check what we got
+		switch len(result) {
+		case 9:
+			return string(result), numberTypeACN
+		case 11:
+			return string(result), numberTypeABN
+		}
+	}
+
+	return "", numberTypeNone
+}
+
+type charType int
```

Comment:
is it better practice to group types/const/etc above?

Reply:
I think usually you just group all of the functionality involving a type in the same area, so when you refactor you would just grab that whole section.